### PR TITLE
Mb/explicit

### PR DESCRIPTION
**Pull request by sesquideus**
Finished converting the `InputSet`s and `Input` classes to the implicit detection of tag parameters (detector, target, ...).
Also added more test to cover for new edge cases (not many though).

### CODE REVIEW

**Comment by chyan26** [comment]
I guess there are issues remains with the simulation data.  Since there is no conflict with current code.  I would suggest we rebase and merge it first.

**Comment by sesquideus** [comment]
Yes, also on my other machine everything works and passes all the tests now... but I do not understand the actual problem.

Still, I think that the benefits of merging now (even without these four tests passing) outweigh the problems it will cause if we diverge too much.

**Comment by hugobuddel** [comment]
I would prefer that we at least identify the problem before merging anything. Could someone else reproduce the problem?

**Comment by sesquideus** [comment]
My bad.
I did not realize I fixed my `pyesorex` ("if not" →  "is None"). The recipes that are failing are exactly those without parameters.

**Comment by hugobuddel** [comment]
Or perhaps we can keep these lines in?
```
    parameters = cpl.ui.ParameterList([])
```

I also want to note that I still had problems even with your fixed pyesorex. Because wouldn't removing this line result in `parameters` to be `None`? As they are set to None in the base class IIRC. So the `is None` would still be trigger, just like the `if not` was, right?

**Comment by sesquideus** [comment]
Yes, we should keep this line in, as a system-wide `MetisRecipe` default.

If I understand it correctly, `cpl.ui.ParameterList([])` is Falsy, and thus it crashes within `pyesorex`. But it is not `None` and that works for me now.

File: metisp/pymetis/src/pymetis/base/recipe.py [hunk 1]
```diff
@@ -27,28 +27,34 @@ class MetisRecipe(cpl.ui.PyRecipe):
...
+    _email = "astar.vienna@univie.ac.at"                        # ToDo is this a sensible default?
+    _copyright = "GPL-3.0-or-later"                             # I guess we're using the same copyright everywhere
     _synopsis = "Abstract-like base class for METIS recipes"
-    _description = "This class serves as the base class for all METIS recipes."
+    _description = ("This class serves as the base class for all METIS recipes."
+                    "Bonus points if it is not visible from pyesorex.")
 
-    parameters = cpl.ui.ParameterList([])          # By default, classes do not have any parameters
-    implementation_class: type = str               # Dummy class, this must be overridden in the derived classes anyway
+    parameters = cpl.ui.ParameterList([])           # By default, a recipe does not have any parameters.
```
Comment (hugobuddel) [inline metisp/pymetis/src/pymetis/base/recipe.py:44]: Maybe add a dummy parameter here? E.g. (untested)
```suggestion
    # By default, a recipe does not have any parameters.
    # However, a default dummy parameter is added because
    # recipes without parameters crash in esorex. See
    # https://jira.eso.org/browse/PIPE-12044
    parameters = cpl.ui.ParameterList([
        cpl.ui.ParameterEnum(
            name="dummy",
            context="dummy",
            description="dummy",
            default="dummy",
            alternatives=("dummy"),
        ),
    ])
```

**Comment by hugobuddel** [comment]
By the way, in my bug report I noted that the first recipe on https://www.eso.org/sci/software/pycpl/pycpl-site/user/basics.html does not specify any parameters, so it actually is `None` there, so that case should be fixed as well (but that is not our problem).

I added a suggestion to add a dummy parameter; would that work for now?

**Comment by hugobuddel** [comment]
I pushed my test branch to our fork as well, see https://github.com/AstarVienna/pyesorex/tree/addtests .

(By the way, I deleted all the nonsense branches in the pyesorex fork, but they keep turning up again, so need to figure out what is happening.)

File: metisp/pymetis/src/pymetis/base/recipe.py (hunk 1 above)
Comment (sesquideus) [reply metisp/pymetis/src/pymetis/base/recipe.py:44]: This will make other tests fail: I verify that the `context` of the parameter is the same as the name of the recipe. But maybe that can be relaxed / deleted.

**Comment by sesquideus** [comment]
Here we go 🥇

File: metisp/pymetis/src/pymetis/base/recipe.py (hunk 1 above)
Comment (hugobuddel) [reply metisp/pymetis/src/pymetis/base/recipe.py:44]: Ah, that's a useful test. Thanks for fixing it like this. A bit ugly, but at least it is clear

**Review by hugobuddel — APPROVED** [review]
It would be good to also add these new recipes to the esorex and EDPS tests, but we can do that later.
